### ISSUE TITLE
`runHook` reports `blocked: false` for a hook that halts the turn with `continue: false`

### ISSUE THREAD

**Issue by vlad-ryzhkov** [zernie/agent-lint#174]
Tested on `vigiles@21.0.1` (npm) = `681edf7`.

`HookOutput` declares the field (`src/run-hook.ts:257`):

```ts
export interface HookOutput {
  readonly decision?: "approve" | "block";
  readonly reason?: string;
  readonly continue?: boolean;      // declared
  readonly stopReason?: string;
  ...
}
```

`decideHook` (`src/run-hook.ts:306`) never reads it:

```ts
  const permission = json?.hookSpecificOutput?.permissionDecision;
  const decision = permission ?? json?.decision;
  const blocked =
    exitCode === protocol.blockExitCode ||
    (decision !== undefined && protocol.denyDecisionValues.includes(decision));
```

`continue: false` is a documented way to stop the turn in Claude Code, and it is stronger than
`permissionDecision: "deny"` — the whole iteration halts and `stopReason` goes back to the agent as
text. Authors pick it exactly when they want to hand back an explanation.

## Measured on a real guard

We have a `PreToolUse` guard that blocks `git push` and `git commit` by printing
`{"continue": false, "stopReason": "…"}`. Through `runHook`:

```
vigiles decideHook: blocked=false decision=undefined — this guard stops the turn
with continue=false, which decideHook does not read
```

So `assertHookBlocked` (`src/harness-assert.ts:126`) fails a test on a hook that does block. We had
to write our own predicate:

```js
const stopsTheTurn = (r) => r.blocked || r.json?.continue === false;
```

## This breaks a flagship feature, not just `runHook`

`verifyGuardrail` (`src/guardrail-check.ts:144-167`) reads `r.blocked` from `runHook`.
`assertBlocksDisasters` (`src/guardrail-check.ts:175`) is line 15 in `what-vigiles-catches.md`: "a
guardrail you wrote that doesn't actually block the disaster battery". Running the battery against
our guard:

```
$ node battery-probe.mjs skills/github-helper/hooks/guard.hook.json
ALLOWED  force-push            (exit 0)
ALLOWED  force-push-compound   (exit 0)
ALLOWED  reset-hard            (exit 0)
ALLOWED  no-verify-commit      (exit 0)
verifyGuardrail says blocked: 0/4
```

Same guard, same four commands, reading `continue`:

```
HALTS  force-push            vigiles.blocked=false continue=false
HALTS  force-push-compound   vigiles.blocked=false continue=false
HALTS  reset-hard            vigiles.blocked=false continue=false
HALTS  no-verify-commit      vigiles.blocked=false continue=false
```

The guard stops all four. `assertBlocksDisasters` would fail the build saying it blocks none. The
tool whose stated job is catching "a guard that looks fine and silently does nothing" reports the
opposite about a working guard.

## Patch — one added line

`src/run-hook.ts`, inside `decideHook`:

```ts
  const blocked =
    exitCode === protocol.blockExitCode ||
    json?.continue === false ||
    (decision !== undefined && protocol.denyDecisionValues.includes(decision));
```

No new field in `HookProtocol` is needed: the comment in `src/core/hook-protocol.ts:1-14` already
says Claude Code and Codex are nearly identical on the wire and the descriptor is deliberately thin.
`continue` is common to both.

If changing the meaning of `blocked` is unwanted, a separate `haltsTurn` on `HookRunResult` gives
the same result without touching existing semantics.

Also worth updating: the doc comment on `HookRunResult.blocked` (`src/run-hook.ts:273-276`) lists
three mechanisms, and a unit test next to the existing `decideHook` ones.

One added line fixes `runHook`, `verifyGuardrail` and `assertBlocksDisasters` at once.

## Adjacent — same patch or a separate one

The `hook-block-ineffective` rule knows the same three mechanisms
(`docs/rules/hook-block-ineffective.md`) and also does not know about `continue`. A
`continue: false` guard wired to an event where it does nothing goes uncaught — a hole in check 1 of
`what-vigiles-catches.md`.

**Comment by zernie** [comment]
Fixed and merged in #186 (`94af23a5`). Thank you — this was the most valuable report of the four, because it was an *inverted verdict* in the feature whose whole job is catching inverted verdicts, and nothing internal had noticed.

**What shipped.** `decideHook` now reads the halt field, so `runHook`, `verifyGuardrail` and `assertBlocksDisasters` all see a halting guard as blocking. Your `stopsTheTurn` workaround predicate is no longer needed.

The one change from your patch: the field is read from the **port**, not written into core.

```ts
const haltField = protocol.haltsTurnField;
const haltsTurn = haltField !== undefined && json?.[haltField] === false;
const blocked =
  exitCode === protocol.blockExitCode ||
  haltsTurn ||
  (decision !== undefined && protocol.denyDecisionValues.includes(decision));
```

You argued no new `HookProtocol` field was needed since the descriptor is thin and `continue` is common to both harnesses. The first half is right and the second is what I could not confirm: `continue` is documented for Claude Code, while `codexHookProtocol`'s notes only record the shared exit-2 / `decision` / `permissionDecision` model. Hard-coding it would have asserted for Codex something nobody measured. So Claude Code declares `haltsTurnField: "continue"` and Codex leaves it unset — a test drives a protocol without the field and asserts `continue: false` is ignored there.

`HookRunResult` also gains `haltsTurn` alongside `blocked`, since a halt is strictly stronger than a deny and worth telling apart.

**Measured, both directions.** A guard halting the turn now passes `assertBlocksDisasters` over the full `DISASTER_CATALOG`; a guard emitting only `continue: true` still **fails** it. Without that second half, "blocked" could be satisfied by the mere presence of a `continue` key — the same bug with the sign flipped. Three mutations, each failing on its own assertion: dropping the halt term, replacing `=== false` with a falsy check, and hard-coding the field past the port.

---

## The adjacent suggestion — deliberately not implemented

You proposed teaching `hook-block-ineffective` the same mechanism. I did the opposite, and you should have the reasoning rather than a silent divergence.

Adding `continue` to that rule's block-attempt set would make it fire on hooks that **work**. Its `wrong-event` check flags a block attempt on `SessionStart` / `SessionEnd` / `Notification` / `PreCompact` — events where a block is silently ignored. But a halt is **not event-scoped**: `continue: false` ends the turn from `SessionStart` just as it does from `PreToolUse`. So every correct halting guard on exactly those events would have been reported as broken. For a rule that can be wired at `error`, a false positive costs the build, and a rule that fails correct builds gets switched off rather than fixed.

What the mechanism does fix there is the reverse: a hook pairing the legacy `"decision":"block"` with a real halt on `PreToolUse` was reported `wrong-field` — "nothing is blocked" — while it blocked. `continue` reads as a **suppressor** of that finding, never as a block attempt.

Both directions are pinned. A mutation that implements your suggestion literally fails on:

```
✗ the halt-the-turn field > a halt on a NO-EFFECT event is NOT flagged wrong-event
```

and removing the suppressor fails on its own case. So if a later reader decides the suppressor reading was wrong, the test says exactly what breaks.

---

**What your report surfaced beyond what it claimed.** The measurement you posted (`0/4` on a working guard) is a cleaner statement of the problem than the type-declaration mismatch you led with. `HookOutput` declaring a field nothing reads is a smell; `assertBlocksDisasters` confidently reporting the opposite about a real guard is the defect. The commit message and the code comment both carry your framing.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by zernie** [comment]
Short follow-up — behaviour unchanged, but the shape your report exposed is gone.

The defect underneath #174 was that `HookOutput` declared a fourth mechanism while `decideHook` was a three-term `||` chain. Fixing the term left the shape: a `||` chain has no form that can be *incomplete*, so nothing would have caught a fifth mechanism arriving the same way.

In #188 (`7693c71`) the mechanisms are a closed table:

```ts
export type BlockMechanism = "exit-code" | "deny-decision" | "halt-field";

const BLOCK_MECHANISMS: Record<BlockMechanism, (ctx: BlockContext) => boolean> = { … };
```

`Record<BlockMechanism, …>` makes a mechanism named-but-unimplemented a tsc error. `HookRunResult` also gains `blockedBy`, so a result says *which* mechanism fired rather than only that one did — and `haltsTurn` is derived from it instead of being a second boolean to keep in sync.

Every mechanism is evaluated, not first-match, so `blocked` and `haltsTurn` compute exactly as they did after #186 — pinned by a test that fails if the table short-circuits, and a mutation dropping an entry fails on `decideHook: continue:false halts the turn`.

Nothing for you to change; posting it because the general form is the part of your report that outlived the specific field.

---
_Generated by [Claude Code](https://claude.ai/code)_
